dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,272 @@
+            bool taskRegistered = std::any_of(begin(tasks), end(tasks),
+            [&](auto&& task)
+            {
+                std::wstring backgroundTaskNameStr = task.Value().Name().c_str();
+                if (backgroundTaskNameStr.find(backgroundTaskName) != std::wstring::npos)
+                {
+                    if (backgroundTaskNameStr.find(taskClsidStr) != std::wstring::npos)
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        throw winrt::hresult_invalid_argument(L"RegisterActivator has different clsid registered.");
+                    }
+                };
+                return false;
+            });
+
+            if (!taskRegistered)
+            {
+                builder = BackgroundTaskBuilder();
+                builder.Name(backgroundTaskName + taskClsidStr.c_str());
+
+                PushNotificationTrigger trigger{};
+                builder.SetTrigger(trigger);
+
+                auto builder5 = builder.try_as<winrt::IBackgroundTaskBuilder5>();
+                if (IsPackagedProcess() && builder5)
+                {
+                    builder5.SetTaskEntryPointClsid(taskClsid);
+                    winrt::com_array<winrt::IBackgroundCondition> conditions = details.GetConditions();
+                    for (auto condition : conditions)
+                    {
+                        builder.AddCondition(condition);
+                    }
+                }
+                else
+                {
+                    throw winrt::hresult_not_implemented();
+                }
+            }
+        }
+
+        if (WI_IsFlagSet(details.Option(), PushNotificationRegistrationOption::ComActivator))
+        {
+            {
+                auto lock = g_lock.lock();
+                // Define handle that will be set during background task execution
+                g_waitHandleForArgs = wil::unique_handle(CreateEvent(nullptr, FALSE, FALSE, nullptr));
+                THROW_HR_IF_NULL(E_UNEXPECTED, g_waitHandleForArgs);
+            }
+
+            THROW_IF_FAILED(::CoRegisterClassObject(
+                taskClsid,
+                winrt::make<PushNotificationBackgroundTaskFactory>().get(),
+                CLSCTX_LOCAL_SERVER,
+                REGCLS_MULTIPLEUSE,
+                &cookie));
+        }
+
+        if (builder)
+        {
+            registeredTask = builder.Register();
+        }
+
+        return PushNotificationRegistrationToken{ cookie, registeredTask };
+    }
+
+    void PushNotificationManager::UnregisterActivator(PushNotificationRegistrationToken const& token, PushNotificationRegistrationOption const& option)
+    {
+        THROW_HR_IF_NULL(E_INVALIDARG, token);
+        if (WI_IsFlagSet(option, PushNotificationRegistrationOption::PushTrigger))
+        {
+            THROW_HR_IF_NULL(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), token.TaskRegistration());
```

Comment:
Suppose somebody calls `RegisterActivator(PushTrigger)` and there is already a background task registered. Therefore, the token that is returned is `{ 0, nullptr }`. They then pass this to `UnregisterActivator` and get `ERROR_NOT_FOUND` because RegisterActivator did not register a task (one already existed). What's the rule for UnregisterActivator? Should it accept anything returned by RegisterActivator? Or are callers expected to inspect it to see if there is a background task inside, and bypass UnregisterActivator if there is no background task?

And suppose the background task was pre-existing. For example, maybe the app crashed after it called RegisterActivator and before it could call UnregisterActivator. How can it get rid of that pre-existing background task?

Reply:
Having a pre-existing background task isn't an issue. It is actually required to allow push notifications to be delivered to the app.

What doesn't seem right to me is to not return the already-registered task, so the app can control when to unregister the background task.